text/0000-mutex_traits.md

```diff
@@ -0,0 +1,239 @@
+the behaviors they implement. However, this kind of 'pivot' cannot be performed on traits by, for example, crate A 
+implementing a trait from std for its new type, and crate B implementing a new trait for all types which implement 
+that same trait.
+
+In addition to the coherence benefits of this change, disallowing a single type to implement two traits has benefits 
+in itself. By encoding mutual exclusivity in the type system, programmers can implement a greater portion of the 
+program logic in a statically analyzable manner. For example, the `num` crate currently defines two traits `Signed` 
+and `Unsigned`. `Unsigned` is a marker trait which exists only to indicate that a type is not `Signed`, but the type 
+system will allow types to implement both `Signed` and `Unsigned` without objection. Clients of `num` cannot actually 
+rely on an `Unsigned` bound to guarantee that a type is not `Signed`, even though that is the trait's only purpose.
+
+# Detailed design
+
+## Trait, !Trait, and ?Trait
+
+An earlier RFC attempted to codify negative reasoning ran aground on the problem of backward compatibility. If it is 
+possible to bound a type parameter or trait by the non-implementation of another trait, then that non-implementation -
+the _absence_ of code - becomes a semantic expression. As an example, consider this system, with traits `Consuamble`, 
+`Edible`, and type `Dirt`.
+
+```rust
+pub trait Consumable {
+    fn consume(&self);
+}
+
+pub trait Edible {
+    fn nourish(&self);
+}
+
+pub type Dirt;
+
+impl<T> Consumable for T where T: Edible {
+    fn consume(&self) { self.nourish() }
+}
+
+impl<T> Consumable for T where T: !Edible {
+    fn consume(&self) {  }
+}
+```
+
+Under the current definitions, consuming `Dirt` has no effect, but if we were to later to discover a way to implement 
+`Edible` for `Dirt`, that would change, and the behavior of any client relying on `Dirt`'s implementation of 
+`Consumable` wouild change.
+
+A solution to this problem which would enable mutual exclusivity is to hold that the relation between types and traits
+is in one of three states for every type and trait, rather than two:
+
+* `T: Trait` - The type `T` implements the trait `Trait`.
+* `T: !Trait` - The type `T` does not implement the trait `Trait`.
+* `T: ?Trait` - The type `T` may or may not implement the trait `Trait`.
+
+Without a contrary expression, the relationship between any `T` and any `Trait` is  `T: ?Trait`. This table documents 
+the three relations and how they are described:
+
+|               | ?Trait            | Trait                | !Trait             |
+|---------------|-------------------|----------------------|--------------------|
+| Specific impl | by default        | impl Trait for T     | impl !Trait for T  |
+|               | impl ?Trait for T |                      |                    |
+| Default impl  | by default        | impl Trait for ..    | impl !Trait for .. |
```

Comment:
Note, that `impl Trait for ..` doesn't currently mean the same thing as `impl<T> Trait for T` (people continue stepping on this rake again and again), it makes `Trait` [very very special] (http://huonw.github.io/blog/2015/05/defaulting-to-thread-safety/#witchcraft) and pretty much unusable for anything except for `Send/Sync`-like things. So either you or OIBIT have to use some other notation.